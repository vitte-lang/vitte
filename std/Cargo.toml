[package]
name        = "vitte-stdlib"
description = "Stdlib Vitte (no_std-first) : IDs typés, Arena, erreurs, utilitaires (bytes/str/time)."

version      = { workspace = true }
edition      = { workspace = true }
rust-version = { workspace = true }
license      = { workspace = true }
repository   = { workspace = true }
homepage     = { workspace = true }
readme       = { workspace = true }

keywords   = ["vitte", "stdlib", "no_std", "arena", "ids"]
categories = ["data-structures", "algorithms", "no-std"]

[lib]
name = "vitte_stdlib"
path = "src/lib.rs"

[features]
# Confort par défaut
default = ["std"]

# Environnements
std        = []   # plein std
alloc-only = []   # no_std + alloc

# Intégrations optionnelles
serde  = ["dep:serde"]
small  = ["dep:smallvec"]
index  = ["dep:indexmap"]
errors = ["dep:thiserror"]
fs     = ["dep:vitte-fs", "std"]
net    = ["dep:vitte-net"]
async  = ["dep:vitte-async"]
# Runtimes async optionnels (se branchent sur vitte-async)
async-rt-tokio      = ["async", "vitte-async/rt-tokio"]
async-rt-async-std  = ["async", "vitte-async/rt-async-std"]
async-rt-smol       = ["async", "vitte-async/rt-smol"]

[dependencies]
# Base (via workspace)
bitflags = { workspace = true }

# Optionnels (via workspace)
serde     = { workspace = true, optional = true, default-features = false, features = ["derive"] }
thiserror = { workspace = true, optional = true }
smallvec  = { workspace = true, optional = true }
indexmap  = { workspace = true, optional = true }
vitte-fs   = { path = "../crates/vitte-fs",   version = "0.1.0", optional = true }
vitte-net  = { path = "../crates/vitte-net",  version = "0.1.0", optional = true }
vitte-async= { path = "../crates/vitte-async",version = "0.1.0", optional = true, default-features = false }

[dev-dependencies]
pretty_assertions = { workspace = true }
proptest          = { workspace = true }
serde_json        = { workspace = true }

[package.metadata.docs.rs]
all-features        = false
features            = ["std", "serde", "small", "index", "errors"]
no-default-features = false
rustdoc-args        = ["--cfg", "docsrs"]

[lints.rust]
unsafe_code = "forbid"

[lints.clippy]
all                     = "warn"
pedantic                = "warn"
nursery                 = "warn"
module_name_repetitions = "allow"
too_many_lines          = "allow"
doc_markdown            = "allow"
