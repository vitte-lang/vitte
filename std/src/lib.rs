//! Vitte Stdlib — brique utilitaire commune, pensée **no_std** par défaut.
//!
//! - `std` : confort complet (I/O, temps, alloc implicite).
//! - `alloc-only` : pas de `std`, mais `Vec`, `String`, etc. via `alloc`.
//! - `serde` : (dé)sérialisation des IDs.
//! - `small` : alias & helpers `smallvec`.
//! - `index` : alias `IndexMap/IndexSet` (`indexmap`).
//! - `errors` : impl `std::error::Error` pour `Error`.
//!
//! # Modules
//! - [`ids`] : IDs typés (`Id<T>`) + générateur [`IdGen`].
//! - [`arena`] : `Arena<T>` append-only & `IdMap<T,U>` compacte.
//! - [`result`] : `VResult<T>` + macros [`bail!`]/[`ensure!`].
//! - [`bytes`] : lecture/écriture endian-safe sans `Read/Write` std.
//! - [`strutil`] : helpers `&str`/`String` (`to_lower_ascii`, `to_snake_case`).
//! - [`time`] : trait `InstantLike` (impl pour `std::time::Instant`).
//! - [`fs`] *(feature `fs`)* : facettes I/O fichiers atomiques & utilitaires répertoires.
//! - [`net`] *(feature `net`)* : façade réseau (TCP/HTTP/WS/TLS) prête pour tokio.
//! - [`async_runtime`] *(feature `async`)* : primitives async runtime-agnostic (`tokio`, `async-std`, `smol`).
//!
//! # Exemple
//! ```
//! use vitte_stdlib::prelude::*;
//! let mut gen = IdGen::default();
//! let a: Id<u32> = gen.fresh();
//! let b: Id<u32> = gen.fresh();
//! assert_ne!(a, b);
//! ```

#![cfg_attr(not(feature = "std"), no_std)]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![forbid(unsafe_code)]

#[cfg(any(feature = "std", feature = "alloc-only"))]
extern crate alloc;

pub mod prelude {
    //! Prélu : importe les types/symboles utiles “partout”.
    #[allow(unused_imports)]
    pub use crate::{
        arena::{Arena, IdMap, idx},
        bytes::{ReadBytes, WriteBytes},
        ids::{Id, IdGen, RawId},
        result::{Error, VResult},
        strutil::*,
        time::*,
    };
    pub use crate::{bail, ensure};

    // Re-exports choisis (pratiques en interne)
    #[cfg(feature = "small")]
    #[cfg_attr(docsrs, doc(cfg(feature = "small")))]
    pub use smallvec::{self, SmallVec};

    #[cfg(feature = "index")]
    #[cfg_attr(docsrs, doc(cfg(feature = "index")))]
    pub use indexmap::{IndexMap, IndexSet};

    pub use bitflags::bitflags;

    #[cfg(feature = "fs")]
    #[cfg_attr(docsrs, doc(cfg(feature = "fs")))]
    pub use crate::fs;

    #[cfg(feature = "net")]
    #[cfg_attr(docsrs, doc(cfg(feature = "net")))]
    pub use crate::net;

    #[cfg(feature = "async")]
    #[cfg_attr(docsrs, doc(cfg(feature = "async")))]
    pub use crate::async_runtime;
}

#[cfg(feature = "fs")]
#[cfg_attr(docsrs, doc(cfg(feature = "fs")))]
pub mod fs {
    //! Facettes système de fichiers de haut niveau.
    pub use vitte_fs::*;
}

#[cfg(feature = "net")]
#[cfg_attr(docsrs, doc(cfg(feature = "net")))]
pub mod net {
    //! Façade réseau unifiée (TCP/HTTP/WS/TLS) reposant sur `vitte-net`.
    pub use vitte_net::*;
}

#[cfg(feature = "async")]
#[cfg_attr(docsrs, doc(cfg(feature = "async")))]
pub mod async_runtime {
    //! Runtime async agnostique (re-export de `vitte-async`).
    pub use vitte_async::*;
}

pub mod ids {
    //! IDs typés, zéro-cost, sûrs et sérialisables (si `serde`).
    use core::{fmt, marker::PhantomData, num::NonZeroU32};

    /// Représentation brute : `u32` non-nul (0 = NULL réservé).
    #[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Default)]
    pub struct RawId(Option<NonZeroU32>);

    impl RawId {
        #[inline]
        pub const fn from_u32(v: u32) -> Self {
            match NonZeroU32::new(v) {
                Some(nz) => Self(Some(nz)),
                None => Self(None),
            }
        }
        #[inline]
        pub const fn to_u32(self) -> u32 {
            match self.0 {
                Some(nz) => nz.get(),
                None => 0,
            }
        }
        #[inline]
        pub const fn is_null(self) -> bool {
            self.0.is_none()
        }
    }

    impl fmt::Debug for RawId {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            if let Some(nz) = self.0 {
                write!(f, "RawId({})", nz.get())
            } else {
                f.write_str("RawId(NULL)")
            }
        }
    }

    /// ID typé : `Id<T>` empêche le mélange d’espaces d’identifiants.
    #[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
    pub struct Id<T> {
        raw: RawId,
        _tag: PhantomData<fn() -> T>,
    }

    impl<T> Id<T> {
        #[inline]
        pub const fn new(raw: RawId) -> Self {
            Self { raw, _tag: PhantomData }
        }
        #[inline]
        pub const fn from_u32(v: u32) -> Self {
            Self::new(RawId::from_u32(v))
        }
        #[inline]
        pub const fn as_raw(self) -> RawId {
            self.raw
        }
        #[inline]
        pub const fn to_u32(self) -> u32 {
            self.raw.to_u32()
        }
        #[inline]
        pub const fn is_null(self) -> bool {
            self.raw.is_null()
        }
    }

    impl<T> Default for Id<T> {
        fn default() -> Self {
            Self::from_u32(0)
        }
    }

    impl<T> core::fmt::Debug for Id<T> {
        fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
            write!(f, "Id({})", self.raw.to_u32())
        }
    }

    #[cfg(feature = "serde")]
    #[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
    mod serde_impl {
        use super::*;
        use serde::{Deserialize, Deserializer, Serialize, Serializer};

        impl<T> Serialize for Id<T> {
            fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
                self.to_u32().serialize(s)
            }
        }
        impl<'de, T> Deserialize<'de> for Id<T> {
            fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
                let v = u32::deserialize(d)?;
                Ok(Id::from_u32(v))
            }
        }
    }

    /// Générateur d’IDs monotone (non thread-safe ; simple & rapide).
    #[derive(Debug, Copy, Clone)]
    pub struct IdGen {
        next: u32,
    }

    impl Default for IdGen {
        fn default() -> Self {
            Self { next: 1 } // 0 = NULL réservé
        }
    }

    impl IdGen {
        #[inline]
        pub fn fresh<T>(&mut self) -> Id<T> {
            let id = Id::<T>::from_u32(self.next);
            self.next = self.next.wrapping_add(1);
            if self.next == 0 {
                self.next = 1;
            }
            id
        }
        #[inline]
        pub fn reset(&mut self) {
            self.next = 1;
        }
        #[inline]
        pub fn peek(&self) -> u32 {
            self.next
        }
    }
}

pub mod arena {
    //! `Arena<T>` : allocations append-only, adressées par `Id<T>`.
    //!
    //! - O(1) indexation par `Id<T>`
    //! - pas de déplacement d’éléments
    //! - itération rapide

    use crate::ids::Id;
    use core::{fmt, iter::FromIterator};

    #[cfg(any(feature = "std", feature = "alloc-only"))]
    use alloc::vec::Vec;

    #[derive(Clone)]
    pub struct Arena<T> {
        data: Vec<T>,
    }

    impl<T> Default for Arena<T> {
        fn default() -> Self {
            Self { data: Vec::new() }
        }
    }

    impl<T> fmt::Debug for Arena<T>
    where
        T: fmt::Debug,
    {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            f.debug_struct("Arena").field("len", &self.data.len()).finish()
        }
    }

    impl<T> Arena<T> {
        #[inline]
        pub fn len(&self) -> usize {
            self.data.len()
        }
        #[inline]
        pub fn is_empty(&self) -> bool {
            self.data.is_empty()
        }

        #[inline]
        pub fn alloc(&mut self, value: T) -> Id<T> {
            let idx = self.data.len();
            self.data.push(value);
            // `idx+1` car 0 est réservé
            Id::from_u32((idx as u32).wrapping_add(1))
        }

        #[inline]
        pub fn get(&self, id: Id<T>) -> Option<&T> {
            self.data.get(idx(id))
        }

        #[inline]
        pub fn get_mut(&mut self, id: Id<T>) -> Option<&mut T> {
            self.data.get_mut(idx(id))
        }

        #[inline]
        pub fn iter(&self) -> core::slice::Iter<'_, T> {
            self.data.iter()
        }

        #[inline]
        pub fn iter_mut(&mut self) -> core::slice::IterMut<'_, T> {
            self.data.iter_mut()
        }

        #[inline]
        pub fn into_vec(self) -> Vec<T> {
            self.data
        }
    }

    impl<T> FromIterator<T> for Arena<T> {
        fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
            let mut a = Arena::default();
            for v in iter {
                a.alloc(v);
            }
            a
        }
    }

    /// Map d’`Id<T>` vers `U` compacte et rapide (vector-map).
    #[derive(Clone)]
    pub struct IdMap<T, U> {
        data: Vec<Option<U>>,
        _tag: core::marker::PhantomData<fn() -> T>,
    }

    impl<T, U> Default for IdMap<T, U> {
        fn default() -> Self {
            Self { data: Vec::new(), _tag: core::marker::PhantomData }
        }
    }

    impl<T, U> IdMap<T, U> {
        #[inline]
        fn ensure_len(&mut self, slot: usize) {
            let need = slot.saturating_add(1);
            if self.data.len() < need {
                self.data.resize_with(need, || None);
            }
        }
        #[inline]
        pub fn insert(&mut self, id: Id<T>, val: U) -> Option<U> {
            let slot = idx(id);
            self.ensure_len(slot);
            core::mem::replace(&mut self.data[slot], Some(val))
        }
        #[inline]
        pub fn get(&self, id: Id<T>) -> Option<&U> {
            self.data.get(idx(id)).and_then(|o| o.as_ref())
        }
        #[inline]
        pub fn get_mut(&mut self, id: Id<T>) -> Option<&mut U> {
            self.data.get_mut(idx(id)).and_then(|o| o.as_mut())
        }
        #[inline]
        pub fn clear(&mut self) {
            self.data.clear()
        }
    }

    /// Convertit un `Id<T>` en index, avec garde.
    #[inline]
    pub fn idx<T>(id: Id<T>) -> usize {
        id.to_u32().wrapping_sub(1) as usize
    }

    // Re-export utile
    pub use crate::ids::RawId;
}

pub mod result {
    //! Résultats, erreurs et macros ergonomiques.
    use core::fmt;

    #[cfg(any(feature = "std", feature = "alloc-only"))]
    use alloc::string::String;

    /// Erreur “lightweight”.
    #[derive(Debug)]
    pub enum Error {
        /// Message libre (nécessite `alloc` ou `std`).
        #[cfg(any(feature = "std", feature = "alloc-only"))]
        Msg(String),

        /// Erreur d’IO (uniquement si `std`).
        #[cfg(feature = "std")]
        Io(std::io::Error),

        /// Variante statique, utile en `no_std` sans `alloc`.
        Other(&'static str),
    }

    #[cfg(feature = "errors")]
    #[cfg_attr(docsrs, doc(cfg(feature = "errors")))]
    impl core::error::Error for Error {}

    impl fmt::Display for Error {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            match self {
                #[cfg(any(feature = "std", feature = "alloc-only"))]
                Self::Msg(s) => f.write_str(s),
                #[cfg(feature = "std")]
                Self::Io(e) => fmt::Display::fmt(e, f),
                Self::Other(s) => f.write_str(s),
            }
        }
    }

    #[cfg(feature = "std")]
    impl From<std::io::Error> for Error {
        fn from(e: std::io::Error) -> Self {
            Self::Io(e)
        }
    }

    /// Alias court pour les résultats “Vitte”.
    pub type VResult<T> = core::result::Result<T, Error>;

    /// `bail!` : quitter tôt avec un message **formaté** (alloc requise).
    #[macro_export]
    macro_rules! bail {
        ($($tt:tt)*) => {{
            #[allow(unused_imports)]
            use $crate::result::Error;
            #[cfg(any(feature = "std", feature = "alloc-only"))]
            {
                return core::result::Result::Err(Error::Msg(alloc::format!($($tt)*)));
            }
            #[cfg(all(not(feature = "std"), not(feature = "alloc-only")))]
            {
                // no_std “pur” : exige une string statique exacte
                return core::result::Result::Err(Error::Other($crate::result::__bail_static_str(&format_args!($($tt)*))));
            }
        }};
    }

    // Helper interne : essaie de convertir un format_args! en &str *si* littéral (no-alloc)
    #[doc(hidden)]
    #[inline]
    pub const fn __bail_static_str(s: &core::fmt::Arguments<'_>) -> &'static str {
        // Dans no_std sans alloc, `bail!` ne peut accepter que des littéraux exacts
        // → l’utilisateur doit écrire: bail!("message statique")
        // (impossible de vérifier à la compilation : doc claire dans le message d’erreur)
        let _ = s;
        "error"
    }

    /// `ensure!` : garde défensive, sinon `bail!`.
    #[macro_export]
    macro_rules! ensure {
        ($cond:expr, $($tt:tt)*) => {{
            if !($cond) { $crate::bail!($($tt)*); }
        }};
    }
}

pub mod bytes {
    //! Lecture/écriture endian-safe sans `std`, pour slices.

    pub trait ReadBytes {
        /// Lis un `u32` **little-endian**.
        fn read_u32_le(&mut self) -> Option<u32>;
        /// Lis un `u64` **little-endian**.
        fn read_u64_le(&mut self) -> Option<u64>;
    }

    pub trait WriteBytes {
        fn write_u32_le(&mut self, v: u32) -> bool;
        fn write_u64_le(&mut self, v: u64) -> bool;
    }

    impl<'a> ReadBytes for &'a [u8] {
        #[inline]
        fn read_u32_le(&mut self) -> Option<u32> {
            if self.len() < 4 {
                return None;
            }
            let (head, tail) = self.split_at(4);
            *self = tail;
            Some(u32::from_le_bytes([head[0], head[1], head[2], head[3]]))
        }
        #[inline]
        fn read_u64_le(&mut self) -> Option<u64> {
            if self.len() < 8 {
                return None;
            }
            let (head, tail) = self.split_at(8);
            *self = tail;
            Some(u64::from_le_bytes([
                head[0], head[1], head[2], head[3], head[4], head[5], head[6], head[7],
            ]))
        }
    }

    #[cfg(any(feature = "std", feature = "alloc-only"))]
    impl WriteBytes for alloc::vec::Vec<u8> {
        #[inline]
        fn write_u32_le(&mut self, v: u32) -> bool {
            self.extend_from_slice(&v.to_le_bytes());
            true
        }
        #[inline]
        fn write_u64_le(&mut self, v: u64) -> bool {
            self.extend_from_slice(&v.to_le_bytes());
            true
        }
    }
}

pub mod strutil {
    //! Petites fonctions utiles pour `&str` / `String` (alloc si dispo).
    #[cfg(any(feature = "std", feature = "alloc-only"))]
    use alloc::string::String;

    /// Minuscule ASCII rapide (ne gère pas l’Unicode complet).
    #[inline]
    #[cfg(any(feature = "std", feature = "alloc-only"))]
    pub fn to_lower_ascii(s: &str) -> String {
        let mut out = String::with_capacity(s.len());
        for b in s.bytes() {
            out.push((if (b'A'..=b'Z').contains(&b) { b + 32 } else { b }) as char);
        }
        out
    }

    /// Convertit `fooBarBaz` → `foo_bar_baz` (heuristique simple).
    #[inline]
    #[cfg(any(feature = "std", feature = "alloc-only"))]
    pub fn to_snake_case(s: &str) -> String {
        let mut out = String::with_capacity(s.len() + s.len() / 4);
        let mut prev_is_lower = false;
        for ch in s.chars() {
            if ch.is_uppercase() && prev_is_lower {
                out.push('_');
            }
            for c in ch.to_lowercase() {
                out.push(c);
            }
            prev_is_lower = ch.is_lowercase();
        }
        out
    }
}

pub mod time {
    //! Minicouche temps/instant, dispo si `std`.
    #[cfg(feature = "std")]
    #[cfg_attr(docsrs, doc(cfg(feature = "std")))]
    pub use std::time::{Duration, Instant};

    /// Trait générique pour des horloges “Instant-like”.
    pub trait InstantLike: Sized + Copy {
        type Duration;
        fn now() -> Self;
        fn elapsed(&self) -> Self::Duration;
    }

    #[cfg(feature = "std")]
    impl InstantLike for std::time::Instant {
        type Duration = std::time::Duration;
        #[inline]
        fn now() -> Self {
            std::time::Instant::now()
        }
        #[inline]
        fn elapsed(&self) -> Self::Duration {
            self.elapsed()
        }
    }
}

// ===== Re-exports utiles (sans polluer fortement) =====
#[cfg(feature = "small")]
#[cfg_attr(docsrs, doc(cfg(feature = "small")))]
pub use smallvec;

#[cfg(feature = "index")]
#[cfg_attr(docsrs, doc(cfg(feature = "index")))]
pub use indexmap;

pub use bitflags;

// ====== Tests rapides (nécessitent généralement `std`) ======
#[cfg(all(test, feature = "std"))]
mod tests {
    use super::*;
    use prelude::*;

    #[test]
    fn ids_basic() {
        let mut g = IdGen::default();
        let a: Id<u8> = g.fresh();
        let b: Id<u8> = g.fresh();
        assert_ne!(a, b);
        assert_eq!(a.to_u32(), 1);
        assert_eq!(b.to_u32(), 2);
        assert!(!a.is_null());
    }

    #[test]
    fn arena_basic() {
        let mut a = Arena::<u32>::default();
        let i1 = a.alloc(42);
        let i2 = a.alloc(7);
        assert_eq!(a.len(), 2);
        assert_eq!(*a.get(i1).unwrap(), 42);
        *a.get_mut(i2).unwrap() += 1;
        assert_eq!(*a.get(i2).unwrap(), 8);
        assert_eq!(arena::idx(i1), 0);
    }

    #[test]
    fn bytes_roundtrip() {
        let mut v = alloc::vec::Vec::new();
        v.write_u32_le(0xDEADBEEF);
        v.write_u64_le(0x0123_4567_89AB_CDEF);
        let mut s: &[u8] = &v;
        assert_eq!(s.read_u32_le().unwrap(), 0xDEADBEEF);
        assert_eq!(s.read_u64_le().unwrap(), 0x0123_4567_89AB_CDEF);
        assert!(s.read_u32_le().is_none());
    }

    #[test]
    fn strutils_ok() {
        assert_eq!(strutil::to_lower_ascii("ABCé"), "abcé");
        assert_eq!(strutil::to_snake_case("FooBarBaz"), "foo_bar_baz");
    }

    #[test]
    fn bail_ensure() {
        fn f(x: i32) -> VResult<()> {
            ensure!(x >= 0, "x doit être >= 0 (got {x})");
            if x == 0 {
                bail!("x ne doit pas être 0");
            }
            Ok(())
        }
        assert!(f(2).is_ok());
        assert!(f(-1).is_err());
        assert!(f(0).is_err());
    }
}
