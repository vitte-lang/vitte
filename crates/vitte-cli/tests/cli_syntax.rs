use std::fs;

use assert_cmd::Command;
use insta::assert_snapshot;
use predicates::prelude::*;
use tempfile::tempdir;

fn base_cmd() -> Command {
    let mut cmd = Command::cargo_bin("vitte").expect("binary `vitte`");
    cmd.env("NO_COLOR", "1");
    cmd.env("CLICOLOR_FORCE", "0");
    cmd
}

fn capture_output(mut cmd: Command) -> Result<(i32, String, String), std::io::Error> {
    let out = cmd.output()?;
    let status = out.status.code().unwrap_or_else(|| if out.status.success() { 0 } else { -1 });
    let stdout = String::from_utf8_lossy(&out.stdout).to_string();
    let stderr = String::from_utf8_lossy(&out.stderr).to_string();
    Ok((status, stdout, stderr))
}

fn snapshot(label: &str, status: i32, stdout: &str, stderr: &str) {
    assert_snapshot!(
        label,
        format!(
            "status: {status}\nstdout:\n{stdout}\
            \n---- stderr ----\n{stderr}"
        )
    );
}

#[test]
fn compile_source_after_options_is_rejected() {
    let mut cmd = base_cmd();
    cmd.args(["compile", "--optimize", "foo.vitte"]);
    let (status, stdout, stderr) = capture_output(cmd).expect("run command");
    snapshot("compile_source_after_options", status, &stdout, &stderr);
}

#[test]
fn run_program_after_options_requires_reordering() {
    let mut cmd = base_cmd();
    cmd.args(["run", "--auto-compile", "main.vitte"]);
    let (status, stdout, stderr) = capture_output(cmd).expect("run command");
    snapshot("run_program_after_options", status, &stdout, &stderr);
}

#[test]
fn global_option_after_command_is_rejected() {
    let mut cmd = base_cmd();
    cmd.args(["compile", "-v"]);
    let (status, stdout, stderr) = capture_output(cmd).expect("run command");
    snapshot("compile_global_option_after_command", status, &stdout, &stderr);
}

fn help_output(args: &[&str]) -> (i32, String, String) {
    let mut cmd = base_cmd();
    cmd.args(args);
    capture_output(cmd).expect("run command")
}

#[test]
fn help_main_command_matches_snapshot() {
    let (status, stdout, stderr) = help_output(&["--help"]);
    snapshot("help_main", status, &stdout, &stderr);
}

#[test]
fn help_compile_command_matches_snapshot() {
    let (status, stdout, stderr) = help_output(&["compile", "--help"]);
    snapshot("help_compile", status, &stdout, &stderr);
}

#[test]
fn help_run_command_matches_snapshot() {
    let (status, stdout, stderr) = help_output(&["run", "--help"]);
    snapshot("help_run", status, &stdout, &stderr);
}

#[test]
fn help_fmt_command_matches_snapshot() {
    let (status, stdout, stderr) = help_output(&["fmt", "--help"]);
    snapshot("help_fmt", status, &stdout, &stderr);
}

#[test]
fn help_inspect_command_matches_snapshot() {
    let (status, stdout, stderr) = help_output(&["inspect", "--help"]);
    snapshot("help_inspect", status, &stdout, &stderr);
}

#[test]
fn help_disasm_command_matches_snapshot() {
    let (status, stdout, stderr) = help_output(&["disasm", "--help"]);
    snapshot("help_disasm", status, &stdout, &stderr);
}

#[test]
fn help_modules_command_matches_snapshot() {
    let (status, stdout, stderr) = help_output(&["modules", "--help"]);
    snapshot("help_modules", status, &stdout, &stderr);
}

#[test]
fn version_flag_prints_version() {
    let mut cmd = base_cmd();
    cmd.args(["--version"]);
    let (status, stdout, stderr) = capture_output(cmd).expect("run command");
    assert_eq!(status, 0);
    assert!(stdout.starts_with("vitte "));
    assert!(stderr.is_empty());
}

#[cfg(feature = "vm")]
#[test]
fn compile_command_writes_bytecode() {
    let dir = tempdir().expect("tempdir");
    let src_path = dir.path().join("main.vitte");
    fs::write(&src_path, "module main;").expect("write source");
    let out_path = dir.path().join("main.vitbc");

    let mut cmd = base_cmd();
    cmd.current_dir(dir.path());
    cmd.args([
        "compile",
        src_path.file_name().unwrap().to_str().unwrap(),
        "--output",
        out_path.file_name().unwrap().to_str().unwrap(),
    ]);

    cmd.assert()
        .success()
        .stdout(predicate::str::is_empty())
        .stderr(predicate::str::contains("COMPILE"));

    assert!(out_path.exists(), "bytecode file should be produced");
}

#[cfg(feature = "fmt")]
#[test]
fn fmt_in_place_success() {
    let dir = tempdir().expect("tempdir");
    let src_path = dir.path().join("format_me.vitte");
    fs::write(&src_path, "module sample;").expect("write source");

    let mut cmd = base_cmd();
    cmd.current_dir(dir.path());
    cmd.args(["fmt", src_path.file_name().unwrap().to_str().unwrap(), "--in-place"]);

    cmd.assert()
        .success()
        .stdout(predicate::str::is_empty())
        .stderr(predicate::str::contains("FMT"));
    assert!(fs::read_to_string(&src_path).unwrap().contains("module sample;"));
}

#[cfg(feature = "fmt")]
#[test]
fn fmt_in_place_and_output_conflict() {
    let mut cmd = base_cmd();
    cmd.args(["fmt", "file.vitte", "--in-place", "--output", "out.vitte"]);
    let (status, stdout, stderr) = capture_output(cmd).expect("run command");
    snapshot("fmt_in_place_output_conflict", status, &stdout, &stderr);
}

#[cfg(feature = "fmt")]
#[test]
fn fmt_in_place_with_stdin_conflict() {
    let mut cmd = base_cmd();
    cmd.args(["fmt", "--in-place"]);
    let (status, stdout, stderr) = capture_output(cmd).expect("run command");
    snapshot("fmt_in_place_stdin_conflict", status, &stdout, &stderr);
}

#[cfg(feature = "fmt")]
#[test]
fn fmt_in_place_with_check_conflict() {
    let mut cmd = base_cmd();
    cmd.args(["fmt", "file.vitte", "--in-place", "--check"]);
    let (status, stdout, stderr) = capture_output(cmd).expect("run command");
    snapshot("fmt_in_place_check_conflict", status, &stdout, &stderr);
}

#[test]
fn compile_invalid_source_reports_parse_error() {
    let dir = tempdir().expect("tempdir");
    let src_path = dir.path().join("bad.vitte");
    fs::write(&src_path, "module main {").expect("write source");

    let mut cmd = base_cmd();
    cmd.current_dir(dir.path());
    cmd.args(["compile", src_path.file_name().unwrap().to_str().unwrap(), "--output", "out.vitbc"]);

    let (status, stdout, stderr) = capture_output(cmd).expect("run command");
    snapshot("compile_invalid_source_parse_error", status, &stdout, &stderr);
}
