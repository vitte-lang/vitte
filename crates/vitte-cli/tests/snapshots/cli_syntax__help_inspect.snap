---
source: crates/vitte-cli/tests/cli_syntax.rs
assertion_line: 24
expression: "format!(\"status: {status}\\nstdout:\\n{stdout}\\\n            \\n---- stderr ----\\n{stderr}\")"
---
status: 0
stdout:
Inspecter un artefact VITBC

Usage: vitte inspect [OPTIONS] [INPUT]

Arguments:
  [INPUT]  Bytecode (.vitbc)

Options:
      --summary            Résumé compact
      --symbols            Table des symboles uniquement
      --consts             Constantes globales
      --hexdump            Dump hexadécimal
      --profile <PROFILE>  Profil d'environnement (dev ou release) [default: dev] [possible values: dev, release]
      --json               Sortie JSON
      --lang <LANG>        Langue pour les messages utilisateur [possible values: en, fr]
      --target             Informations cible
      --header             Informations d'en-tête
      --sections           Liste des sections
      --strings            Chaînes ASCII embarquées
      --imports            Fonctions importées
      --exports            Fonctions exportées
      --disasm             Désassemblage textuel
      --debug              Informations de debug
      --meta               Métadonnées additionnelles
      --verify             Vérification d'intégrité
      --size               Taille par section
      --deps               Dépendances embarquées
      --entry              Point d'entrée
      --dump-all           Active toutes les sections
  -h, --help               Print help

---- stderr ----
